### PR TITLE
Build Pyodide into workerd

### PR DESCRIPTION
**Pull request by hoodmane**
This uses bazelisk to download the upstream Pyodide v0.25.0a1 release and compile it into workerd. We should soon make our own custom build of Pyodide because upstream Pyodide loads slowly. But this is a start.

### CODE REVIEW

File: src/pyodide/patch_pyodide_js.py
Comment (dom96) [inline src/pyodide/patch_pyodide_js.py:1]: Could this script just be a diff that is applied via `git apply`? (like we already do for v8 patches)

Non-blocking I think but might be worth it to get rid of this script.

File: src/workerd/api/pyodide/pyodide.h
Comment (dom96) [inline src/workerd/api/pyodide/pyodide.h:12]: Check your autogate here and do nothing if it's not enabled.

File: src/pyodide/python.js
Comment (dom96) [inline src/pyodide/python.js:1]: Non-blocking: could/should a lot of the code in here be upstreamed to Pyodide? of course the imports are workerd-specific, but could the snapshot logic/filesystem prep be upstreamed?

**Review by dom96 — APPROVED** [review]
Left some comments, but in general LGTM.

File: src/pyodide/patch_pyodide_js.py
Comment (hoodmane) [reply src/pyodide/patch_pyodide_js.py:1]: I don't want to make a patch to prepend stuff to a file then we'd have to include context of the beginning of the file. The beginning of the file might change but we don't care. The point of a patch file is if you do care about matching the context.

Comment (hoodmane) [reply src/pyodide/patch_pyodide_js.py:1]: Bazel does require a lot of boilerplate to do it this way though...

File: src/pyodide/python.js
Comment (hoodmane) [reply src/pyodide/python.js:1]: The filesystem stuff is already present upstream, but it's awkward for us to use it because it's bundled into `pyodide.js` so we'd have to use a custom build. I'm thinking about upstreaming the memory snapshotting, but it'll probably take a while for me to get around to it.

File: src/pyodide/internal/builtin_wrappers.js
Comment (hoodmane) [inline src/pyodide/internal/builtin_wrappers.js:18]: @dom96 see here for eval safety check

File: src/workerd/api/pyodide/eval.h
Comment (hoodmane) [reply src/workerd/api/pyodide/eval.h:10]: I think this is not the best place for it since it has to be imported by an internal builtin JS module. So any internal module that imports it should be responsible for making sure they don't leak it improperly. I added an extra callee check in `builtin_wrappers.js` [see here](https://github.com/cloudflare/workerd/pull/1467/files#r1414626194).

File: src/pyodide/internal/builtin_wrappers.js
Comment (hoodmane) [reply src/pyodide/internal/builtin_wrappers.js:18]: I guess to make sure that this actually works, we'd have to do a security review to check that it's not possible for a user to make a module whose name renders as `pyodide-internal:pyodide-bundle/pyodide.asm`...

Comment (hoodmane) [reply src/pyodide/internal/builtin_wrappers.js:18]: Well the point of this is that I'm concerned that global scope symbols could somehow leak from `pyodide.asm.js`. This `WebAssembly` patch is itself a global symbol. I don't *think* there's anyway for user code to get to it, but it's best to be cautious. OTOH, if we permit calling the function on anyone who has a symbol in global scope it kind of feels like the same check again. But perhaps from C++ we have some stronger way to validate the identity of `pyodide.asm.js`.

But I think this check as-is is less fragile than you might think. Someone would have to actually make a file called `pyodide-internal:pyodide-bundle/pyodide.asm.js`, but I don't think users are allowed to make files with names like this. Since I split the colons from the right, even if someone managed to make a function with a name like `createFunction (pyodide-internal:...)`, we would still notice the correct parentheses.

Comment (hoodmane) [reply src/pyodide/internal/builtin_wrappers.js:18]: OTOH, I wonder if they could spoof it by overriding `Error.prepareStackTrace`. We should consider setting and restoring `Error.prepareStackTrace` here and then we wouldn't have to parse the traceback.

Comment (hoodmane) [reply src/pyodide/internal/builtin_wrappers.js:18]: Okay, I made an improved version with `prepareStackTrace`.

I think the best thing to do would be to have an actual list of function objects that are allowed to call the function, and then add a C++ function that returns the function up the callstack (forbidden in ordinary strict mode JS) so we can call that to look up the callee and check if it's in a whitelist. That check would be impossible to spoof.

File: src/workerd/api/pyodide/eval.h
Comment (jasnell) [reply src/workerd/api/pyodide/eval.h:10]: Rather than this, I'd prefer an approach where this exports a function like...

```
impl.eval(() => { /** ... **/ });
```

Where the `js.setAllowEval(false)` is automatically handled when the call to `impl.eval(...)` completes.

File: src/pyodide/internal/builtin_wrappers.js
Comment (jasnell) [inline src/pyodide/internal/builtin_wrappers.js:1]: Note that we already have the `UnsafeEval` object (see `api/unsafe.h`). It is currently used only in a binding but could also be exposed by the `pyodide-internal:eval` import here.

File: src/pyodide/internal/builtin_wrappers.js
Comment (jasnell) [inline src/pyodide/internal/builtin_wrappers.js:60]: This is rather unfortunate from a perf point of view. `checkCallee()` is creating an Error just to get the stack, `Reflect.construct...` has some performance overhead, etc. I have to wonder if it wouldn't be more efficient just to implement this entire piece in c++... not sure tho. We'll have to play with it a bit.

File: src/workerd/api/pyodide/pyodide.h
Comment (jasnell) [inline src/workerd/api/pyodide/pyodide.h:11]: Should we also gate this on the `--experimental` compat flag? at least for now.

File: src/workerd/api/pyodide/eval.h
Comment (jasnell) [inline src/workerd/api/pyodide/eval.h:23]: I think we can replace this entirely with `UnsafeEval` ... but if we don't, this should be `EW_PYODIDE_EVAL_ISOLATE_TYPES`

File: src/pyodide/internal/builtin_wrappers.js
Comment (hoodmane) [reply src/pyodide/internal/builtin_wrappers.js:1]: I'm not sure what you mean. As a guess are you suggesting adding `UnsafeEval::newWebAssemblyModule`? That seems reasonable.

File: src/pyodide/internal/builtin_wrappers.js
Comment (hoodmane) [reply src/pyodide/internal/builtin_wrappers.js:16]: I forget about this, but it can be pretty dramatic...

File: src/pyodide/internal/builtin_wrappers.js
Comment (hoodmane) [reply src/pyodide/internal/builtin_wrappers.js]: Okay I did this.

File: src/workerd/api/pyodide/pyodide.h
Comment (hoodmane) [reply src/workerd/api/pyodide/pyodide.h]: Sounds good to me/

**Comment by dom96** [comment]
Looks like the CI is failing because autogates are not initialised in the `TestFixture`. Here is a diff that fixes that:

```
diff --git a/src/workerd/tests/test-fixture.c++ b/src/workerd/tests/test-fixture.c++
index f5ab2e53..8b5a6677 100644
--- a/src/workerd/tests/test-fixture.c++
+++ b/src/workerd/tests/test-fixture.c++
@@ -189,6 +189,14 @@ inline server::config::Worker::Reader buildConfig(
   auto modules = config.initModules(1);
   modules[0].setName(mainModuleName);
   modules[0].setEsModule(params.mainModuleSource.orDefault(mainModuleSource));
+
+  // Initialise autogates with an empty config. TODO(later): allow TestFixture to accept autogate
+  // states and pass them in here.
+  //
+  // This needs to happen here because `buildConfig` is called early in the construction of
+  // `TestFixture`.
+  util::Autogate::initAutogate({});
+
   return config;
 }
```

File: src/workerd/api/unsafe.h
Comment (jasnell) [inline src/workerd/api/unsafe.h:53]: You can pass the `src` in here as a `kj::Array<kj::byte>`. The type wrapper system will automatically handle the conversion. It'll save you a few lines in your impl.

File: src/workerd/api/unsafe.h
Comment (jasnell) [inline src/workerd/api/unsafe.h:67]: Nit: just being pedantic. Let's call this `internal:unsafe-eval` just to make it extra clear there's something special about it.

File: src/workerd/api/pyodide/pyodide.h
Comment (jasnell) [inline src/workerd/api/pyodide/pyodide.h:14]:
```suggestion
  if (featureFlags.getWorkerdExperimental() &&
      util::Autogate::isEnabled(util::AutogateKey::BUILTIN_WASM_MODULES)) {
    registry.addBuiltinBundle(PYODIDE_BUNDLE, kj::none);
  }
```

File: src/workerd/api/modules.h
Comment (jasnell) [inline src/workerd/api/modules.h:21]:
```suggestion
  registerUnsafeModules(registry, featureFlags);
```

File: src/pyodide/internal/builtin_wrappers.js
Comment (jasnell) [inline src/pyodide/internal/builtin_wrappers.js:60]: Why console.warn here as opposed to propagating a proper error?

File: src/pyodide/patch_pyodide_js.py
Comment (jasnell) [reply src/pyodide/patch_pyodide_js.py:1]: At the very least, a comment at the top of this file that describes its purpose and how it fits in overall would be super helpful for future generations :-)

File: src/pyodide/python.js
Comment (jasnell) [inline src/pyodide/python.js:60]: This is using the alternative Proxies `WebAssembly` correct? I'd generally prefer that we use a different name for it.

File: samples/pyodide/worker.py
Comment (jasnell) [inline samples/pyodide/worker.py:4]: Nit: calling this just `fetch` would align more with the JS worker API.

File: src/workerd/api/unsafe.h
Comment (hoodmane) [reply src/workerd/api/unsafe.h:53]: Presumably that also makes it work for a general `BufferSource`?

File: build/wd_js_bundle.bzl
Comment (mikea) [inline build/wd_js_bundle.bzl:118]: let's update the comment.

I highly recommend bazel vscode plugin which will highlight buildifier issues.

File: build/wd_ts_bundle.bzl
Comment (mikea) [inline build/wd_ts_bundle.bzl:43]: same, comment.

File: build/wd_ts_bundle.bzl
Comment (mikea) [inline build/wd_ts_bundle.bzl:28]: what's the idea here? If you don't need typescript use wd_js_bundle, otherwise I don't see apoint.

File: samples/pyodide/config.capnp
Comment (mikea) [inline samples/pyodide/config.capnp:1]: let's add readme to the dir.

File: samples/pyodide/config.capnp
Comment (mikea) [inline samples/pyodide/config.capnp:17]: add a comment that this enables the experimental feature it depends upon.

File: samples/pyodide/worker.mjs
Comment (mikea) [inline samples/pyodide/worker.mjs:4]: it is not very clear to me how much here is boiler plate and will need to be customized and why? Or will this file be in reality copy-pasted by everyone?

File: src/pyodide/BUILD.bazel
Comment (mikea) [inline src/pyodide/BUILD.bazel:6]: need to format the file.

**Review by mikea — CHANGES_REQUESTED** [review]
I highly suggest separating dep+build change from actually making pyodide to work.
This will make it easier to review and land considering it has downstream implication.

Otherwise `ts = False` is my biggest gripe here and needs to be removed.

File: src/pyodide/BUILD.bazel
Comment (hoodmane) [reply src/pyodide/BUILD.bazel:6]: What do we use for starlark code formatting? I ran black on this.

File: build/wd_ts_bundle.bzl
Comment (hoodmane) [reply build/wd_ts_bundle.bzl:28]: I think the transformations done in  the `wd_js_bundle` invocation here (lines 61 through 108 or so) should be moved into the `wd_js_bundle` command and then we can just pass the arguments here unaltered and I can call `wd_js_bundle` directly.

**Comment by dom96** [comment]
> I highly suggest separating dep+build change from actually making pyodide to work. This will make it easier to review and land considering it has downstream implication.

I'd appreciate it if we could move forward instead of splitting this into even more PRs (this is already nr. 3 in a series of PRs and we have other PRs depending on this). We need to get this in so we can iterate on it in our internal repo (right now I'm blocked on this).

Build changes should be very low risk, so I'd rather we merge in what we have and iterate on the build files in parallel rather than block this PR further.

File: samples/pyodide/worker.mjs
Comment (hoodmane) [reply samples/pyodide/worker.mjs:4]: Yeah, the plan is to generate whatever boilerplate is appropriate from config in followup work. This PR is just supposed to setup the implementation for that, and this `sample` is intended as an indication of how it works right now.

**Review by jasnell — APPROVED** [review]

**Review by dom96 — APPROVED** [review]

File: src/workerd/api/pyodide/pyodide.h
Comment (mikea) [inline src/workerd/api/pyodide/pyodide.h:9]: maybeRegisterPyodideModules?

File: src/pyodide/python.js
Comment (mikea) [inline src/pyodide/python.js:51]: Rather than documenting something as a simplified version of external code I'd prefer a documentation of what does _this_ function do.

File: src/pyodide/python.js
Comment (mikea) [inline src/pyodide/python.js:68]: sounds like important piece of configuration. let's make it top-level const?

File: src/pyodide/python.js
Comment (mikea) [inline src/pyodide/python.js:106]: let's add high-level documentation of the steps this function performs.

**Review by mikea — CHANGES_REQUESTED** [review]

File: src/workerd/api/modules.h
Comment (hoodmane) [reply src/workerd/api/modules.h:21]: This makes `UnsafeEval` available only to builtin/internal modules. So I don't think this change by itself is risky: by itself it makes `UnsafeEval` accessible in various builtin things but none of them use it so it's dead code. OTOH we are using it somewhere so we have to be careful about that.

I can gate it on experimental too.

Comment (mikea) [reply src/workerd/api/modules.h:21]: I think a mere existence of `UnsafeEval` was controversial. I'd like @jasnell to chime in: is it ok to always register UnsafeEval?

> I can gate it on experimental too.

I think it would be best if you installed it only if python modules are installed (which is already gated?)

File: src/pyodide/BUILD.bazel
Comment (hoodmane) [reply src/pyodide/BUILD.bazel:25]: I renamed the rules as suggested. But I think putting it in this subfolder makes it clearer in `python.js` that the imports are generated files. What about putting them into a folder called `generated/`?

**Comment by jasnell** [comment]
> Unconditionally installing unsafe eval for everyone is definitely a big security concern.

This is not unconditionally installing it for everyone. This exposes it only through an internal-only import that is only available to built-in modules.

File: src/workerd/api/modules.h
Comment (jasnell) [reply src/workerd/api/modules.h:21]: > Does this make the unsafe eval module usable by all workers?

No, as long as it is registered as `INTERNAL` it is only accessible by our own built-ins.

> I think a mere existence of `UnsafeEval` was controversial

Not so much. It just shouldn't be accessible to regular worker code, which it is not.

File: src/pyodide/internal/builtin_wrappers.js
Comment (jasnell) [reply src/pyodide/internal/builtin_wrappers.js:45]: v8 requires eval to be enabled when compiling and evaluating the WASM module. We temporarily switch eval on when we compile and evaluate WASM modules in general (see `compileWasmGlobal()` in workerd-api).

File: src/workerd/api/unsafe.c++
Comment (mikea) [inline src/workerd/api/unsafe.c++:97]: should we reuse compileWasmGlobal here? It also has something about background threads in it.

Comment (jasnell) [reply src/workerd/api/unsafe.c++:97]: The `compileWasmGlobal` method is currently defined in `workerd-api.c++` for workerd. The internal repo has its own copy of the function. We could likely move that into jsg and use the one impl in every case. I wouldn't block this PR on it tho.

File: src/pyodide/internal/builtin_wrappers.js
Comment (mikea) [reply src/pyodide/internal/builtin_wrappers.js:45]: Ah I see. UnsafeEval is the last place though, where I'd look for compileWasmGlobal(). Is it to much to ask to extract it to something like `WasmSupport` module instead?

Comment (jasnell) [reply src/pyodide/internal/builtin_wrappers.js:45]: I'll work on consolidating the wasm compile into a utility in jsg that can be used here and by both the external workerd-api and the internal equivalent. But that can be done separately and doesn't need to block here.

File: src/workerd/api/unsafe.c++
Comment (mikea) [reply src/workerd/api/unsafe.c++:97]: how about `AllowV8BackgroundThreadsScope`? Do we need it here?

File: src/pyodide/internal/builtin_wrappers.js
Comment (hoodmane) [reply src/pyodide/internal/builtin_wrappers.js:45]: The same V8 gate for eval also controls wasm compilation. It makes sense because both invoke code generation.

**Comment by hoodmane** [comment]
Perhaps the `UnsafeEval` changes should also be split off then? Probably it would also be good to add a basic test for `newWasmModule` to `unsafe-test.js`.

File: src/workerd/api/unsafe.c++
Comment (jasnell) [reply src/workerd/api/unsafe.c++:97]: We probably do, yes.

**Comment by hoodmane** [comment]
Okay I added some basic tests for `newWasmModule` in `unsafe-test.js` and a separate test that we can't import `UnsafeEval` from user code directly to `src/workerd/server/tests/unsafe-eval/`.

File: src/workerd/api/modules.h
Comment (hoodmane) [reply src/workerd/api/modules.h:21]: Okay I added some tests for this.

File: src/pyodide/patch_pyodide_js.py
Comment (mikea) [reply src/pyodide/patch_pyodide_js.py:41]: I think sed will be
- less to maintain
- impossible to be clever during patching (which I consider a good thing).

Comment (hoodmane) [reply src/pyodide/patch_pyodide_js.py]: Okay I switched to using `expand_template` as you suggested.

File: src/pyodide/patch_pyodide_js.bzl
Comment (mikea) [inline src/pyodide/patch_pyodide_js.bzl:56]: this could be a simple macro, the rule is not necessary. sorry :)

File: src/pyodide/python.js
Comment (mikea) [inline src/pyodide/python.js:1]: documentation here is much, much better now. Thanks a lot!

**Review by mikea — APPROVED** [review]

File: src/pyodide/patch_pyodide_js.bzl
Comment (hoodmane) [reply src/pyodide/patch_pyodide_js.bzl:56]: I tried using `expand_template` directly and I got a name error. Do macros get access to `ctx.actions`? Or how do they get to `expand_template`?

Comment (mikea) [reply src/pyodide/patch_pyodide_js.bzl:56]: you just 

`load("@bazel_skylib//rules:expand_template.bzl", "expand_template")`

and use `expand_template()` directly as a rule in your macro. Or are you talking about something else?

Comment (hoodmane) [reply src/pyodide/patch_pyodide_js.bzl:56]: It seems that the expand template rule was first added in `bazel_skylib` v1.3.0:
https://github.com/bazelbuild/bazel-skylib/commit/2a87d4a62af886fb320883aba102255aba87275e

We're using `bazel_skylib` v1.2.1 currently. But I guess it should be easy enough to update.

**Pull request closed without merging**
